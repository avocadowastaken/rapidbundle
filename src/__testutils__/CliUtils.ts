import { promises as fs } from "fs";
import * as path from "path";
import { bundle } from "../index";
import { PackageJSON } from "../manifests/NodePackageManifest";
import snapshotDiff = require("snapshot-diff");

interface FixtureOutputs {
  main?: string;
}

const builtCode = new Set<string>();
function trackOutputs(outputs: FixtureOutputs): void {
  for (const code of Object.values(outputs)) {
    builtCode.add(code);
  }
}

expect.addSnapshotSerializer({
  test(value: unknown) {
    return typeof value == "string" && builtCode.has(value);
  },
  print(value: unknown) {
    return (value as string).trim();
  },
});

export async function buildFixture(
  code: string,
  pkg: PackageJSON = {}
): Promise<FixtureOutputs> {
  const { testPath, currentTestName } = expect.getState();
  const name = `${path
    .basename(testPath)
    .replace(/\./g, "_")}__${currentTestName}`;

  const pkgDir = path.join(process.cwd(), "node_modules", ".fixtures", name);

  await fs.rm(pkgDir, { force: true, recursive: true });

  const srcDir = path.join(pkgDir, "src");

  await fs.mkdir(srcDir, { recursive: true });

  await fs.writeFile(path.join(srcDir, "index.ts"), code, "utf8");
  await fs.writeFile(
    path.join(pkgDir, "package.json"),
    JSON.stringify({ name, ...pkg }),
    "utf8"
  );

  await bundle(pkgDir);

  const outputs: FixtureOutputs = {};

  if (pkg.main) {
    outputs.main = await fs.readFile(path.join(pkgDir, pkg.main), "utf8");
  }

  trackOutputs(outputs);

  return outputs;
}

export async function injectModule(name: string, code: string) {
  const pkgDir = path.join(process.cwd(), "node_modules", name);

  await fs.rm(pkgDir, { force: true, recursive: true });
  await fs.mkdir(pkgDir, { recursive: true });

  await fs.writeFile(path.join(pkgDir, "index.js"), code, "utf8");
  await fs.writeFile(
    path.join(pkgDir, "package.json"),
    JSON.stringify({
      name,
      exports: { ".": "./index.js" },
    }),
    "utf8"
  );
}
