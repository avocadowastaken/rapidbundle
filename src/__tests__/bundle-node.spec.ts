import { buildFixture, injectModule } from "../__testutils__/CliUtils";

beforeAll(async () => {
  await injectModule(
    "is-function",
    `
export function isFunction(input) {
  return typeof input == 'function';
}
`
  );
});

test("basic", async () => {
  const { main } = await buildFixture(
    `
import { isFunction } from 'is-function';
export const CAN_USE_DOM = typeof document == 'object' && isFunction(document?.createElement);
`,
    { main: "dist.js" }
  );

  expect(main).toMatchInlineSnapshot(`
    var __defProp = Object.defineProperty;
    var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
    var __name = (target, value) => __defProp(target, "name", {value, configurable: true});
    var __export = (target, all) => {
      for (var name in all)
        __defProp(target, name, {get: all[name], enumerable: true});
    };

    // node_modules/.fixtures/bundle-node_spec_ts__basic/src/index.ts
    __markAsModule(exports);
    __export(exports, {
      CAN_USE_DOM: () => CAN_USE_DOM
    });

    // node_modules/is-function/index.js
    function isFunction(input) {
      return typeof input == "function";
    }
    __name(isFunction, "isFunction");

    // node_modules/.fixtures/bundle-node_spec_ts__basic/src/index.ts
    var CAN_USE_DOM = typeof document == "object" && isFunction(document == null ? void 0 : document.createElement);
    // Annotate the CommonJS export names for ESM import in node:
    0 && (module.exports = {
      CAN_USE_DOM
    });
  `);
});

test("version", async () => {
  const { main } = await buildFixture("export let foo = bar?.baz;", {
    main: "dist.js",
    engines: { node: ">=10" },
  });

  expect(main).toMatchInlineSnapshot(`
    var __defProp = Object.defineProperty;
    var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
    var __export = (target, all) => {
      for (var name in all)
        __defProp(target, name, {get: all[name], enumerable: true});
    };

    // node_modules/.fixtures/bundle-node_spec_ts__version/src/index.ts
    __markAsModule(exports);
    __export(exports, {
      foo: () => foo
    });
    var foo = bar == null ? void 0 : bar.baz;
    // Annotate the CommonJS export names for ESM import in node:
    0 && (module.exports = {
      foo
    });
  `);
});

test("dependencies", async () => {
  const { main } = await buildFixture(
    `
import { isFunction } from 'is-function';
import debug = require('debug');
import isObject = require('isobject');
export const logger = debug('app');
if (isObject(console) && isFunction(console.log)) logger.log = console.log.bind(console);
`,
    {
      main: "dist.js",
      dependencies: { debug: "*" },
      peerDependencies: { isobject: "*" },
    }
  );

  expect(main).toMatchInlineSnapshot(`
    var __defProp = Object.defineProperty;
    var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
    var __name = (target, value) => __defProp(target, "name", {value, configurable: true});
    var __export = (target, all) => {
      for (var name in all)
        __defProp(target, name, {get: all[name], enumerable: true});
    };

    // node_modules/.fixtures/bundle-node_spec_ts__dependencies/src/index.ts
    __markAsModule(exports);
    __export(exports, {
      logger: () => logger
    });

    // node_modules/is-function/index.js
    function isFunction(input) {
      return typeof input == "function";
    }
    __name(isFunction, "isFunction");

    // node_modules/.fixtures/bundle-node_spec_ts__dependencies/src/index.ts
    var debug = require("debug");
    var isObject = require("isobject");
    var logger = debug("app");
    if (isObject(console) && isFunction(console.log))
      logger.log = console.log.bind(console);
    // Annotate the CommonJS export names for ESM import in node:
    0 && (module.exports = {
      logger
    });
  `);
});
