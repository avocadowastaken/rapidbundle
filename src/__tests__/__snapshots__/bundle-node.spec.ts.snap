// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __name = (target, value) => __defProp(target, "name", {value, configurable: true});
var __commonJS = (cb, mod) => () => (mod || cb((mod = {exports: {}}).exports, mod), mod.exports);
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// node_modules/isobject/index.js
var require_isobject = __commonJS((exports, module2) => {
  "use strict";
  module2.exports = /* @__PURE__ */ __name(function isObject2(val) {
    return val != null && typeof val === "object" && Array.isArray(val) === false;
  }, "isObject");
});

// node_modules/__fixtures__/bundle-node-basic/src/index.ts
__markAsModule(exports);
__export(exports, {
  isAdmin: () => isAdmin
});
var import_isobject = __toModule(require_isobject());
function isAdmin(user) {
  return (0, import_isobject.default)(user == null ? void 0 : user.role);
}
__name(isAdmin, "isAdmin");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isAdmin
});
/*!
 * isobject <https://github.com/jonschlinkert/isobject>
 *
 * Copyright (c) 2014-2017, Jon Schlinkert.
 * Released under the MIT License.
 */

`;

exports[`version 1`] = `
Object {
  "main": "Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
  var import_isobject = __toModule(require_isobject());
  function isAdmin(user) {
-   return (0, import_isobject.default)(user == null ? void 0 : user.role);
+   return (0, import_isobject.default)(user?.role);
  }
  __name(isAdmin, \\"isAdmin\\");",
}
`;
